src/plugins/embeddable/public/lib/containers/container.ts

```diff
@@ -132,6 +132,54 @@ export abstract class Container<
     return this.createAndSaveEmbeddable(type, panelState);
   }
 
+  /**
+   * This can be used if replacing an emeddable has different consquences than adding one -
+   * for example, adding a new control embeddable will cause re-ordering but replacing an
+   * existing control embeddable will not
+   */
+  protected getReplacementPanelState<
+    EEI extends EmbeddableInput = EmbeddableInput,
+    EEO extends EmbeddableOutput = EmbeddableOutput,
+    E extends IEmbeddable<EEI, EEO> = IEmbeddable<EEI, EEO>
+  >(
+    id: string,
+    factory: EmbeddableFactory<EEI, EEO, E>,
+    newExplicitInput: Partial<EEI>
+  ): PanelState<EEI> {
+    newExplicitInput.id = id;
+    const newPanelState = this.createNewPanelState<EEI, E>(factory, newExplicitInput);
+    return newPanelState;
+  }
+
+  public async replaceEmbeddable<
+    EEI extends EmbeddableInput = EmbeddableInput,
+    EEO extends EmbeddableOutput = EmbeddableOutput,
+    E extends IEmbeddable<EEI, EEO> = IEmbeddable<EEI, EEO>
+  >(id: string, newExplicitInput: Partial<EEI>, newType?: string) {
+    if (!this.input.panels[id]) {
+      throw new PanelNotFoundError();
+    }
+
+    if (newType && newType !== this.input.panels[id].type) {
+      const factory = this.getFactory(newType) as EmbeddableFactory<EEI, EEO, E> | undefined;
+      if (!factory) {
+        throw new EmbeddableFactoryNotFoundError(newType);
+      }
+
+      const newPanelState = this.getReplacementPanelState(id, factory, newExplicitInput);
+      this.updateInput({
+        panels: {
+          ...this.input.panels,
+          [id]: newPanelState,
```

Comment:
As I mentioned in an earlier comment, this could be simplified by spreading all old panel state and only replacing the type and input. This would mean that regardless of the type of the container's panel state, all the rest of it (order, gridData etc) would be retained. Something like:

``` typescript
...this.input.panels,
[id]: {
  ...this.input.panels[id],
  explicitInput: newExplicitInput,
  type: newType,
},
```

If this sort of code works properly we could remove the `getReplacementPanelState` method entirely.

Reply:
Yup, you're totally right! Tried it out and this works great 👍